korma/src/commonMain/kotlin/com/soywiz/korma/geom/Vector3D.kt

```diff
@@ -5,95 +5,190 @@ import com.soywiz.korma.interpolation.interpolate
-        fun length(x: Float, y: Float, z: Float): Float = sqrt(lengthSq(x, y, z))
-
-        fun lengthSq(x: Double, y: Double, z: Double, w: Double): Double = x * x + y * y + z * z + w * w
-        fun lengthSq(x: Double, y: Double, z: Double): Double = x * x + y * y + z * z
-        fun lengthSq(x: Float, y: Float, z: Float, w: Float): Float = x * x + y * y + z * z + w * w
-        fun lengthSq(x: Float, y: Float, z: Float): Float = x * x + y * y + z * z
+    fun copyInto(other: Vector3D): Vector3D {
+        data.copyInto(other.data)
+        return other
     }
 
-    fun copyFrom(other: Vector3D) = setTo(other.x, other.y, other.z, other.w)
+    fun copy() = Vector3D(x, y, z)
 
-    fun setTo(x: Float, y: Float, z: Float, w: Float): Vector3D = this.apply { this.x = x; this.y = y; this.z = z; this.w = w }
-    fun setTo(x: Double, y: Double, z: Double, w: Double): Vector3D = setTo(x.toFloat(), y.toFloat(), z.toFloat(), w.toFloat())
-    fun setTo(x: Int, y: Int, z: Int, w: Int): Vector3D = setTo(x.toFloat(), y.toFloat(), z.toFloat(), w.toFloat())
+    inline fun setTo(func: (index: Int) -> Float): Vector3D = setTo(func(0), func(1), func(2))
 
-    fun setTo(x: Float, y: Float, z: Float): Vector3D = setTo(x, y, z, 1f)
-    fun setTo(x: Double, y: Double, z: Double): Vector3D = setTo(x, y, z, 1.0)
-    fun setTo(x: Int, y: Int, z: Int): Vector3D = setTo(x, y, z, 1)
+    fun setTo(x: Float, y: Float, z: Float): Vector3D {
+        this.x = x
+        this.y = y
+        this.z = z
+        return this
+    }
 
-    inline fun setToFunc(func: (index: Int) -> Float): Vector3D = setTo(func(0), func(1), func(2), func(3))
-    inline fun setToFunc(l: Vector3D, r: Vector3D, func: (l: Float, r: Float) -> Float) = setTo(
+    inline fun setTo(l: Vector3D, r: Vector3D, func: (l: Float, r: Float) -> Float) = setTo(
         func(l.x, r.x),
         func(l.y, r.y),
-        func(l.z, r.z),
-        func(l.w, r.w)
+        func(l.z, r.z)
+    )
+
+    fun setToInterpolated(left: Vector3D, right: Vector3D, t: Float) = setTo(
+        t.interpolate(left.x, right.x),
+        t.interpolate(left.x, right.x),
+        t.interpolate(left.x, right.x)
     )
-    fun setToInterpolated(left: Vector3D, right: Vector3D, t: Double): Vector3D = setToFunc { t.interpolate(left[it], right[it]) }
 
-    fun scale(scale: Float) = this.setTo(this.x * scale, this.y * scale, this.z * scale, this.w * scale)
-    fun scale(scale: Int) = scale(scale.toFloat())
-    fun scale(scale: Double) = scale(scale.toFloat())
+    fun scale(scale: Float) = setTo(x * scale, y * scale, z * scale)
 
     fun transform(mat: Matrix3D) = mat.transform(this, this)
 
-    fun normalize(vector: Vector3D = this): Vector3D = this.apply {
-        val norm = 1.0 / vector.length3
-        setTo(vector.x * norm, vector.y * norm, vector.z * norm, 1.0)
+    /**
+     * Normalizes this object and returns the length before normalization.
```

Comment:
Do you have other APIs as reference doing this to determine if this is a common practice?
With chainable, you could do: `normalize().length` that better states what you are doing.
And if this is an optimization, maybe a method called `normalizeAndReturnLength()` would state better the intent of this function